corrected transition_bias estimation, fixes #55

DonkeyShot21 opened the pull request:
This pull request fixes issue #55.

What is in this pull request:

- new function `uisrnn.utils.estimate_transition_bias` that calculates transition_bias correctly
- removed transition_bias estimation snippet from `uisrnn.utils.resize_sequence`. Now resize_sequence only retuns `sub_sequences` and `seq_lengths`
- moved transition_bias update (for multiple calls to `fit`) from `fit_concatenated` to `fit`
- added warning message if `train_sequences` is already concatenated and `transition_bias` has not been passed as argument, saying that the estimation can be incorrect
- edited few test scripts to adapt to the new shape of the data returned from `uisrnn.utils.resize_sequence`

I also wanted to add `_` to `fit_concatenated` to make it private so that the user is discouraged to call it, but I didn't because maybe it has some other dependency to some internal code of yours.

wq2012 commented on tests/utils_test.py:
```diff
@@ -136,7 +136,7 @@ def test_resize_sequence_without_permutation2(self):
+    sub_sequence, seq_lengths = utils.resize_sequence(
         sequence=np.array([[1, 1], [2, 2], [3, 3]]),
         cluster_id=np.array([1, 2, 1]),
         num_permutations=2)
```
Would you mind adding some simple tests for the estimate_transition_bias() function you added?

DonkeyShot21 replied:
Yes, I can. It will take some days though.

wq2012 replied:
No worries.

wq2012 commented on uisrnn/utils.py:
```diff
@@ -310,3 +300,25 @@ def output_result(model_args, training_args, test_record):
+
+  Args:
+    cluster_id: A list of cluster indicator sequences, must be a
+      non-concatenated sequence.
```
Instead of say "must be", what about this:
cluster_id: Either a list of cluster indicator sequences, or a single concatenated sequence. The former is strongly preferred, since the transition_bias estimated from the latter will be inaccurate.

DonkeyShot21 replied:
OK, sure!

DonkeyShot21 commented on uisrnn/utils.py:
```diff
@@ -310,3 +300,25 @@ def output_result(model_args, training_args, test_record):
+  Args:
+    cluster_id: A list of cluster indicator sequences, must be a
+      non-concatenated sequence.
+
```
Sorry, I forgot to add docstring. I chose to make the default value 1 because that was the value it had previously in your code (you didn't call it smooth, just added 1). It seemed you put it to avoid -inf value in np.log in the case a sequence with a single speaker is provided. I also thought it was a good idea to add it to the denominator as well to avoid division by 0. Using a smaller value would decrease the bias on the calculation of transition_bias but can also lead to underflow in some very remote cases. I can design some tests that prove my point. 
Do you have any suggestion on the value of `smooth`?

wq2012 replied:
I think `smooth=1` is good. Thanks. Please add the docstring to reflect what you explain in this comment.

DonkeyShot21 commented on uisrnn/utils.py:
```diff
@@ -310,3 +300,25 @@ def output_result(model_args, training_args, test_record):
+
+
+def estimate_transition_bias(cluster_ids, smooth=1):
+  """estimate transition bias.
```
OK!

DonkeyShot21 commented:
Ok, sorry for the inconvenience.

DonkeyShot21 commented:
Ok, it took some time but I think we are good now.
This is what I have added:
- reverted all changes to docs
- fixed the docstring
- now the denominator takes double the smoothing of the numerator. This prevents both `np.log(transition_bias)`, and `np.log(1 - transition_bias)` to go to `-inf`. Previous smoothing only worked with `np.log(transition_bias)`
- simple test for the new function.
